text/0000-native-cli.md

```diff
@@ -0,0 +1,136 @@
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## Creating a new Neon project
+
+Neon provides templates for conveniently creating various types of Neon projects, including:
+
+- libraries;
+- Node apps; and
+- Electron apps.
+
+### Libraries
+
+To create a new Neon library, use the command:
+
+```
+npm init neon-lib my-lib
+```
+
+Use whatever name you want for your library in place of `my-lib`.
+
+### Node apps
+
+To create a new Node app, use the command:
+
+```
+npm init neon-app my-app
+```
+
+Use whatever name you want for your library in place of `my-app`.
+
+### Electron apps
+
+To create a new Electron app, use the command:
+
+```
+npm init neon-electron-app my-app
+```
+
+Use whatever name you want for your library in place of `my-app`.
+
+## Building a Neon project
+
+All of the standard Neon project templates are set up to allow you to do a complete build from source with the `install` script, meaning that all you have to do to build your project is the standard `npm install` (or `npm i` for short) command from the project root.
+
+If you look at your `package.json`, you'll see that the install script delegates to Rust's `cargo` tool to build the native addon from Rust source. You can do this directly yourself by running `cargo build` (for a debug build) or `cargo build --release` (for a release build) from the project root.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Replacing `neon new`
+
+The reason we can now move past `neon new` is that `npm` has added support for custom templates via extensibility of the `npm init` command. We have pre-reserved npm packages called:
+
+- `create-neon-lib`
+- `create-neon-app`
+- `create-neon-electron-app`
+
+The extensible design of [`npm init`](https://docs.npmjs.com/cli/init) means that these packages are automatically used as project generators for the commands:
+
+- `npm init neon-lib <args...>`
+- `npm init neon-app <args...>`
+- `npm init neon-electron-app <args...>`
+
+respectively.
+
+## Replacing `neon build`
+
+There are a couple of key developments that will allow us to move past `neon build`:
+
+1. Once we move to the [N-API backend](https://github.com/neon-bindings/neon/issues/444), we no longer need to invalidate a Neon build when changing between versions of Node. This means we won't need to use the custom invalidation logic (which has proved to be fragile and a source of user frustration).
+2. We discovered that we can skip the final step of copying the built DLL into a `.node` file by replacing the JS boilerplate that loads the addons with a call to the lower-level but official [`process.dlopen()`](https://nodejs.org/api/process.html#process_process_dlopen_module_filename_flags) API.
```

Comment:
Unfortunately, `process.dlopen()` may not be usable. Cargo build scripts are intentionally kept from knowing the location of the `target` directory. This makes it difficult for our generated `index.js` to find the lib file.

Instead, we can pass a `rustc` flag to output the library in the correct location. This is possible because build scripts _can_ know the location of the project directory.

```rust
// in build.rs
let manifest_dir = env::var_os("CARGO_MANIFEST_DIR").unwrap();
let output = Path::new(&manifest_dir).join("index.node");

println!("cargo:rustc-cdylib-link-arg=-o");
println!("cargo:rustc-cdylib-link-arg={}", output.display());
```

This sample code has been tested and works, but would need to be improved with platform dependent code to name the file correctly in Windows.

Reply:
Incorporated in the latest draft.